accepted/2021/InboxSourceGenerators.md

```diff
@@ -0,0 +1,165 @@
+
+Source generators should not generate code which uses public API with "reserved" or "undocumented" functionality in attempts to avoid public API guarantees.  Libraries paired with source generators need to provide the same guarantees as libraries alone do today.
+
+Source generators should avoid emitting public API in a user's projects.  Such API can create a compatibility contract in that user's library which needs to be maintained.  If a source generator produces public API in a user's project by design, that public API must be strictly maintained following the same rules as framework components across versions of the source generator.  Care should be taken to the public API which is generated and should be reviewed by the Framework Design Council.  A similar constraint applies to internal API: users will have source that expects to interact with the generated internal API and the user will expect that code to continue to compile between versions.
+
+Generated code should not be generated in System namespaces, it should prefer the user's namespace.  Generated code should be triple-slashed commented and be free from compiler warnings.  Generated code should avoid name collisions with user code and provide affordances for resolving name conflicts.  As much as possible generated code should meet the quality statndards of .NET library code, using nullable annotations meeting framework design and naming guidelines, etc.  It should feel like a natural extension of the library.
+
+### Source generator servicing
+
+Source generators contribute significant code to user assemblies and that code may have issues.  We have very little control over changing customer assemblies once they have consumed a source generator.  We cannot and should not attempt to change the behavior of customer's assemblies (eg: hot patch) to fix issues.  Should an issue be found we should fix it in the source generator and ship a new version, just as we would if we found an issue with compiler.  Customer's desiring this fix must proactively acquire the new code-generator, typically through an updated SDK which carries the ref pack, rebuild their code, and update their production applications.
+
+Due to the limited servicing agility for code-generated code, we should be very careful about the complexity of the code which is generate.  Where possible we should limit code-generated code to "glue code", only that which must be specific to the user assembly or types.  We should try to put more complex code inside the framework or library itself, even if it means exposing public API that is specifically for source generated code to call.
+
+**Open issue:** should we do anything to make it easier to identify assemblies which might need an update?  We can consider some case studies.  
```

Comment:
```
$"[GeneratedCodeAttribute(
    \"{GetType().AssemblyQualifiedName}\",
    \"{Assembly.ExecutingAssembly.MoreWordsHere.FileVersion}\")]"
```

perhaps?

Reply:
This isn't a technical question, it's a policy/experience.  The technical side of it is less interesting and should follow from policy side.

Reply:
I'll use your suggestion in the list of options/examples lower in the doc, it's better than AssemblyMetadataAttribute.